File: boat-sample/src/test/java/sample/java/xyz/srclab/common/state/StateSample.java
package sample.java.xyz.srclab.common.state;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.testng.annotations.Test;
import xyz.srclab.common.state.State;
import xyz.srclab.common.test.TestLogger;

public class StateSample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testState() {
        MyState myState = new MyState(1, "description");
        MyState newState = myState.withMoreDescription("cause");
        //description[cause]
        logger.log(newState.description());
    }

    public static class MyState implements State<Integer, String, MyState> {

        private final int code;
        private final String description;

        public MyState(int code, String description) {
            this.code = code;
            this.description = description;
        }

        @Override
        public Integer code() {
            return code;
        }

        @Nullable
        @Override
        public String description() {
            return description;
        }

        @NotNull
        @Override
        public MyState withNewDescription(@Nullable String newDescription) {
            return new MyState(code, newDescription);
        }

        @NotNull
        @Override
        public MyState withMoreDescription(@Nullable String moreDescription) {
            return new MyState(code, State.moreDescription(description, moreDescription));
        }
    }
}

File: boat-sample/src/test/kotlin/sample/kotlin/xyz/srclab/common/state/StateSample.kt
package sample.kotlin.xyz.srclab.common.state

import org.testng.annotations.Test
import xyz.srclab.common.state.CharsState
import xyz.srclab.common.state.CharsState.Companion.joinStateDescriptions
import xyz.srclab.common.state.CharsState.Companion.moreDescriptions
import xyz.srclab.common.state.State
import xyz.srclab.common.test.TestLogger

class StateSample {

    @Test
    fun testState() {
        val myState = MyState(1, "description")
        val newState = myState.withMoreDescription("cause")
        //description[cause]
        logger.log(newState.description)
    }

    class MyState(
        override val code: Int, override val descriptions: List<String>
    ) : State<Int, String, MyState> {

        constructor(code: Int, description: String?) : this(code, CharsState.newDescriptions(description))

        override val description: String? = descriptions.joinStateDescriptions()

        override fun withNewDescription(newDescription: String?): MyState {
            return MyState(code, CharsState.newDescriptions(newDescription))
        }

        override fun withMoreDescription(moreDescription: String): MyState {
            return MyState(code, descriptions.moreDescriptions(moreDescription))
        }
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}